File: p98.cpp
#include <bits/stdc++.h>
#include <TreeNode.hpp>

using namespace std;

class Solution {
public:
    bool traverse(TreeNode* root, long low, long high) {
        if(root == NULL)
            return true;
        if(root->val <= low || root->val >= high)
            return false;
        bool left = traverse(root->left, low, root->val);
        bool right = traverse(root->right, root->val, high);
        return left && right;
    }
    bool isValidBST(TreeNode* root) {
        return traverse(root, LONG_MIN, LONG_MAX);
    }
};

File: p98.py
from DS.TreeNode import TreeNode

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:

        def validate(root, lo=float('-inf'), hi=float('inf')):
            if root is None:
                return True
            if root.val <= lo or root.val >= hi:
                return False
            return validate(root.left, lo, root.val) and validate(root.right, root.val, hi)

        return validate(root)
